functions/src/api/authenticateJWT.ts

```diff
@@ -0,0 +1,46 @@
+import * as jsonWebToken from 'jsonwebtoken'
+import { Request, Response } from 'express'
+import { VerifyErrors } from 'jsonwebtoken'
+import { JWTData } from '../types/JWTData'
+import { JWTToken } from '../types/JWTToken'
+import { isTokenValidInDb } from '../db/isTokenValidInDb'
+import { assertJWTEnv } from '../utils/assertConfig'
+
+export const authenticateJWT = (
+    req: Request,
+    res: Response,
+    next: Function
+) => {
+    const jwtKey = assertJWTEnv()
+
+    const authHeader = req.headers.authorization
+
+    if (authHeader) {
+        const token = <JWTToken>authHeader.split(' ')[1]
+
+        jsonWebToken.verify(
+            token,
+            jwtKey,
+            {
+                algorithms: ['HS256'],
+            },
+            async (err: VerifyErrors | null, data: {} | undefined) => {
+                if (err || !data) {
+                    return res.sendStatus(403)
+                }
+                const jwtData: JWTData = <JWTData>data
+
+                const isTokenValid = await isTokenValidInDb(jwtData, token)
+
+                if (!isTokenValid) {
+                    return res.sendStatus(403)
+                }
```

Comment:
please consider keeping a single failure exit point for this function:

```typescript
try {
   // when error occurs:
   throw new Error(...)
   // success case

} catch(e) {
   // error case
   return res.sendStatus(403)
}
```

this removes the duplication and increases readability

Reply:
how would you suggest refactoringthe method?

Reply:
even more 
```
async (err: VerifyErrors | null, data: {} | undefined) => {
  try {
    if (err || !data) {
      throw new ForbiddenError() // new type to define here
    }

    const jwtData: JWTData = <JWTData>data
    await assertTokenValidInDb(jwtData, token) // this can raise a ForbiddenError
    next()
  } catch (e) {
   next(e)
  }
}
```

and use an error middleware
```
errorMiddleware((error: Error, req, res, next) => {
  res.sendStatus(error.status)
})

app.use(errorMiddleware)
```